### ISSUE TITLE
Arm64 support

### ISSUE THREAD

**Issue by throwaway876** [phpipam-docker/phpipam-docker#3]
Apologies- didn't see any tickets here when I posted so didn't think this portion was the place for it. This is a copy of https://github.com/phpipam/phpipam/issues/3184 for tracking. To bring it fully here- I was able to build the base for arm64 easily enough and the www and cron ones would just have to build off that with the same specs. Build scripts are welcome to share if you can and what you build against (ie. a machine with docker or hopefully one that can be specced to use docker buildx, etc.) 

No rush as I too have things going on but I wouldn't mind helping improve the process if we can find something that works.
Thanks Gary

Original below
-------------------------------------------

**Is your feature request related to a problem? Please describe.**
Somewhat- just cannot run the image on a Pi 4.

**Describe the solution you'd like**
Build the image with arm64 support. I need to test this image in particular but with docker's buildx it's been easy for me to convert others and can be easily implemented into a pipeline that runs on a docker builder with buildx support enabled.

**Describe alternatives you've considered**
Limiting where this pod can run to x86_64 nodes only.

**Describe how the feature would benefit other users**
Would allow people running K8s and K3s on smaller boards (ie. Pi, etc.) to run without issue.

**Comment by GaryAllan** [comment]
Hi,

The containers are built via a bash script `./build.container master nightly` and `./build.container 1.4 latest 1.4x`

Cutdown version below....

``` bash
#!/bin/bash

BRANCH=$1  
TAG1=$2
TAG2=$3

cd $BASE/repos/phpipam.$BRANCH
git pull

BUILD_DATE=`date -u +"%Y-%m-%dT%H:%M:%SZ"`
VCS_REF=`git rev-parse --short HEAD`
LABELS="--build-arg BUILD_DATE=$BUILD_DATE --build-arg VCS_REF=$VCS_REF --build-arg VERSION=$TAG1 --build-arg BRANCH=$BRANCH"

echo "**** Building WWW image"
cd $BASE/dockerfiles/phpipam-www

if [ "$TAG2" != "" ]; then
  docker build $LABELS --no-cache -t phpipam/phpipam-www:$TAG1 -t phpipam/phpipam-www:$TAG2 .
else
  docker build $LABELS --no-cache -t phpipam/phpipam-www:$TAG1 .
fi

echo "**** Building CRON image"
cd $BASE/dockerfiles/phpipam-cron

if [ "$TAG2" != "" ]; then
  docker build $LABELS --no-cache -t phpipam/phpipam-cron:$TAG1 -t phpipam/phpipam-cron:$TAG2 .
else
  docker build $LABELS --no-cache -t phpipam/phpipam-cron:$TAG1 .
fi

```

I've started researching docker buildx. I have access to a 64bit Pi.  It's currently running 32bit Rasbian and is performing other duties. I want to avoid rebuilding it if possible. 

The VM performing the builds is running binfmt-support 2.1.6 and according to https://nexus.eddiesinentropy.net/2020/01/12/Building-Multi-architecture-Docker-Images-With-Buildx/ binfmt-support 2.1.8 is required for docker buildx with QEMU.

I may have to upgrade the build VM OS unless there is another way to get a working multi-arch build environment.

**Comment by throwaway876** [comment]
I've used this site with great effect for upgrading DRBD piecemeal (as needed for newer kernels on some things I manage) - https://pkgs.org/download/binfmt-support 

I don't know if just upgrading the binfrm_support package by itself will be fine or not (or if it'll need dependencies along with that upgrade.) I do know that I tried building on Ubuntu 16.04 VMs I have (in my current cluster) and ran into the issues you saw (I wish I had found the link you sent as it would've saved me some time.) I went and built a new Ubuntu 20.04 LTS VM, added docker-ce 19.03, docker-ce-cli, default binfmt_support and qemu-user-support (or qemu-user, can't remember), enabled it in the docker config and it worked to build. 

I would be happy to pull arm64 builds (and perhaps armv6 which I think is what other Pis run on if I can find an old Pi around here) and test. Test case of one is at least better than none. I'll just set a node selector in my K8s cluster to schedule that deployment on the Pi(s).

**Comment by GaryAllan** [comment]
Hi

phpIPAM has a 64bit php requirement so no armv6/v7.

I've upgraded the VM to Ubuntu 20.04LTS so the binfrm_support issues are also resolved.

The next issue is docker buildx won't --load images when multipe platforms are set (somthing to do with the docker client and manifests).  I need to --push builds to a repository.

I didn't want to make phpipam-base public on Dockerhub, it's just a base OS image for other containers. I'm looking at private repositories. Portus looks promising....

I'll post when something is ready for testing.

**Comment by GaryAllan** [comment]
Hi @throwaway876 
I've uploaded ARM64 images to Dockerhub. Are you able to test?

**Comment by throwaway876** [comment]
Hey @GaryAllan , just tested. I updated my test K8s deployment with it (along with the node selector and tolerations) and it works just fine. Was able to drop right in with my existing database/data. Verified all is good on existing.

Also just did a quick docker run on the Pi as well and it seems just fine too (not that it shouldn't be.) The new install screen pulls up like normal, etc. I'll let you know if I have any issues but so far seems like it's doing just fine. I have it set to always pull the image so I can maybe make some kind of testing pipeline to pull weekly and do some smoke tests or something or other just to verify but honestly I would assume it'd be just fine.

I appreciate you getting on it so quickly.
